### PR TITLE
Leaves - Dianna

### PR DESCRIPTION
**Pull request by dtingg**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a method that gets the top ten works in a specific category, sorted by number of votes (descending).
Describe how you approached testing that model method. What edge cases did you come up with? | I tried to test all possible scenarios.  I tested categories that had more than 10 works, less than 10 works, and 0 works.
What are `session` and `flash`? What is the difference between them? | Session is a special hash that we can use to track data related to a user’s session, which ends when they close their browser. Flash is a special hash that we can use to send one-time messages from our controllers to our views.  Flash messages only last until the next request-response cycle, whereas session data is stored indefinitely.
What was one thing that you gained more clarity on through this assignment? | I definitely learned more about working with Active Record and creating complex queries.
What is the Heroku URL of your deployed application? | https://dianna-media-ranker.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? | On the main site, I couldn’t recognize any logic for breaking ties.  I think it would be helpful to add some kind of secondary sorting criteria.  For example, on a user’s details page, you should sort the votes by title or creation date.

### CODE REVIEW

File: app/views/works/_work_list.html.erb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      <tr>
+        <th width=80>Votes</th>
+        <th>Title</th>
+        <th>Created By</th>
+        <th width=100>Published</th>
+        <th width=100>Upvote</th>
+      </tr>
+    </thead>
+    <tbody>
+        <% @works.where(category: work_name).each do |work| %>
```
Comment (dHelmgren) [inline app/views/works/_work_list.html.erb:14]: Nice work on the view partial

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<h2>List of Works</h2>
+<%= render partial: "work_list", locals: { work_name: "album" } %>
+<%= render partial: "work_list", locals: { work_name: "book" } %>
+<%= render partial: "work_list", locals: { work_name: "movie" } %>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:4]: 👩‍🍳 👌

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | Yes, but no message
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | yes
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Great work overall! I don't really have any suggestions!
